File: aoc2023/day_1.py
from functools import reduce

# part one
with open("./data/day_1.txt") as fin:
    counter = 0
    for line in fin:
        digits = (int(c) for c in line if c.isdigit())
        first = last = next(digits)
        for last in digits:
            pass
        counter += first * 10 + last
counter

# part two
numbers = (
    "one",
    "two",
    "three",
    "four",
    "five",
    "six",
    "seven",
    "eight",
    "nine",
)
counter = 0.0
with open("./data/day_1.txt") as fin:
    counter = 0
    for line in fin:
        numstrs = []
        for ii, c in enumerate(line):
            if line[ii].isdigit():
                numstrs.append(int(line[ii]))
            else:
                for jj, number in enumerate(numbers, 1):
                    if line[ii : ii + len(number)] == number:
                        numstrs.append(jj)
                        break
        counter += 10 * numstrs[0] + numstrs[-1]
print(counter)

File: rusty2023/src/bin/day1.rs
use std::fs;
use std::cmp;
use std::io::{prelude::*, BufReader};

pub fn part_one() -> u32 {
    let file = fs::File::open("./data/day1.txt");
    let reader = BufReader::new(file.unwrap());

    let sum = reader
        .lines()
        .map(|line| {
            let iter = line
                .expect("something bad happened")
                .chars()
                .filter(|c| c.is_numeric())
                .map(|x| x.to_digit(10).unwrap())
                .collect::<Vec<u32>>();
            return iter.first().unwrap() * 10u32 + iter.last().unwrap();
        })
        .sum::<u32>();
    return sum;
}

pub fn part_two() -> u32 {
    let numbers = vec![
        "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
    ];

    let file = fs::File::open("./data/day1.txt");
    let reader = BufReader::new(file.unwrap());

    let sum = reader
        .lines()
        .map(|line| {
            let iter = line
                .as_ref()
                .expect("something bad happened")
                .chars()
                .enumerate()
                .map(|(ii, c)| {
                    if c.is_digit(10) {
                        return c.to_digit(10).unwrap();
                    } else {
                        return numbers
                            .iter()
                            .enumerate()
                            .filter(|(jj, number)| {
                                number.to_string()
                                    == line.as_ref().expect("something")
                                        [ii..cmp::min(ii + number.chars().count(), line.as_ref().expect("something").chars().count())]
                            })
                            .map(|(jj, number)| return u32::try_from(jj).unwrap() + 1)
                            .next()
                            .unwrap_or(0u32);
                    }
                })
                .filter(|x| x != &0u32)
                .collect::<Vec<u32>>();
            return iter.first().unwrap() * 10u32 + iter.last().unwrap();
        })
        .sum();
    return sum;
}

fn main() {
    println!("{}", part_one());
    println!("{}", part_two());
}
